collector/collector.cpp

```diff
@@ -81,19 +83,85 @@ ShutdownHandler(int signum) {
   g_control.store(ControlValue::STOP_COLLECTOR);
 }
 
+static void WriteBuffer(char* buffer, int num_bytes, int buffer_size) {
+  if (num_bytes > buffer_size) {
+    // the output was truncated, to keep it readable add a new line
+    buffer[buffer_size - 1] = '\n';
+    write(STDERR_FILENO, buffer, buffer_size);
+  } else {
+    write(STDERR_FILENO, buffer, num_bytes);
+  }
+}
+
 static void AbortHandler(int signum) {
-  // Write a stacktrace to stderr
-  void* buffer[32];
-  size_t size = backtrace(buffer, 32);
-  backtrace_symbols_fd(buffer, size, STDERR_FILENO);
+  // Write a stacktrace to stderr. Since the abort handler could be called
+  // after SIGSEGV or some other emergency situations:
+  // * we have to be modest and use little memory (in case if the original
+  //   reason was OOM)
+  // * as the original implementation noted, use reentrant functions
+  //
+  // The latter one was refering to snprintf, although it's not cleat if it is
+  // reentrant. Libc docs says functions using I/O streams are potentially
+  // non-reentrant [1], and it seems under the hood snprintf uses streams.
+  //
+  // [1]: https://www.gnu.org/savannah-checkouts/gnu/libc/manual/html_node/Nonreentrancy.html
+
+  void* buffer[32];  // Addresses buffer
+
+  char message_buffer[256];  // Actual text buffer for a single stack trace
+                             // address. Some templated signatures could be
+                             // particularly long.
+  int num_bytes;
+
+  size_t n_frames = backtrace(buffer, 32);
+  size_t max_frames = sizeof(buffer) / sizeof(buffer[0]);
+  int message_buffer_size = sizeof(message_buffer);
+
+  for (size_t i = 1; i < n_frames; i++) {
```

Comment:
Why do we start from 1 and not 0?

Reply:
Somehow I was under the impression, that the first address is not what we're looking for, but let me check.

Reply:
Now I remember, the first function would be the signal handler, which is not necessarily what we're looking for. But at the same time doesn't hurt to report it as well.